---
title: "qq_plots_of_discrete_distributions_on_transcriptomic_data"
author: "Mangiola Stefano"
date: "31/01/2019"
output: html_document
---

```{r setup, include=FALSE, echo=FALSE}
# Import libraries

#setwd("/wehisan/bioinf/bioinf-data/Papenfuss_lab/projects/mangiola.s/PostDoc/RNAseq-noise-model")
set.seed(13254)

# Import libraries
library(tidyverse)
library(magrittr)
library(rstan)
library(tidybayes)
library(here)
library(foreach)
library(doParallel)
registerDoParallel()
# library(future)
# plan(multiprocess)

rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())

#devtools::load_all() # Sorry this costs me 30 minutes every day

my_theme = 	
	theme_bw() +
	theme(
		panel.border = element_blank(),
		axis.line = element_line(),
		panel.grid.major = element_line(size = 0.2),
		panel.grid.minor = element_line(size = 0.1),
		text = element_text(size=12),
		legend.position="bottom",
		aspect.ratio=1,
		axis.text.x = element_text(angle = 90, hjust = 1),
		strip.background = element_blank(),
		axis.title.x  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
		axis.title.y  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
	)

# Tools
source("https://gist.githubusercontent.com/stemangiola/90a528038b8c52b21f9cfa6bb186d583/raw/b7862745e36ad7a7d2250e0c49ee36dbabc36164/transcription_tool_kit.R")

PIG_model =
  rstan::stan_model(
    here::here("stan",sprintf("%s.stan", "poisson_inverseGaussian"))
  )

```

# T cells most heavy-tailed genes

```{r load_data, cache=TRUE}
counts = read_csv("t_cells_with_error_qq_plot_means.csv")
```

```{r predict_values, cache=TRUE, fig.dim = c(9, 9)}

	counts %>%
  inner_join((.) %>% distinct(symbol, `gene error mean`) %>% arrange(`gene error mean` %>% desc) %>% head(n=25)) %>%
  arrange(`gene error mean` %>% desc) %>%
  mutate(symbol = factor(symbol, unique(symbol))) %>%

  #filter(symbol=="FOSB") %>%
  
  # Infer NB
  group_by(symbol) %>%
   do({
 
     `%>%` = magrittr::`%>%`
		library(tidyverse)
		library(magrittr)
     # lines(seq(0,1000, 10) , sapply(seq(0,1000, 10)  , gamlss.dist::dSICHEL,mu= 100, sigma=0.05, nu=-0.3, log=T))
    #browser()
    # Negative binomial
    fit_NB =  (.) %>%
       pull(`read count normalised`) %>%
      MASS::fitdistr("Negative Binomial")
    
    #PIG
    fit_PIG = 
      optimizing(
        PIG_model, 
        data = list(N = (.) %>% nrow, y = (.) %>% pull(`read count normalised`)),
        init = list(mu = (.) %>% pull(`read count normalised`) %>% `+` (1) %>% log %>% mean)
        ) %$% par
    
    # # Sichel
    # fit_SH =  (.) %>%
    #    arrange(`read count`) %>%
    #    pull(`read count`) %>%
    #   MASS::fitdistr(gamlss.dist::dSICHEL, list(mu= 100, sigma=0.1, nu=-5), upper = c(4000, 20, 0),  lower=c(0, 0.0001, -10))

    
    (.) %>%
      arrange(`read count normalised`) %>%
      mutate(
      	predicted_NB = 
      		qnbinom(
	      		ppoints(`read count normalised`), 
	      		size=(fit_NB[1] %$% estimate)[1], 
	      		mu=(fit_NB[1] %$% estimate)[2]
	      		)
      ) %>%
      # PIG
      mutate(
      	predicted_PIG = 
      		gamlss.dist::qPIG(
      			ppoints(`read count normalised`), 
      			mu = fit_PIG[2] %>% exp, 
      			sigma=fit_PIG[1]
      			
      	) 
      )
    #%>%
      # mutate(
      # 	predicted_SH = 
      # 		gamlss.dist::qSICHEL(
      # 			ppoints(`read count`), 
      # 			nu = (fit_SH[1] %$% estimate)[3], 
      # 			sigma=(fit_SH[1] %$% estimate)[2], 
      # 			mu=(fit_SH[1] %$% estimate)[1], 
      # 			max.value = max(`read count`)
      # 			)
      # 	) %>%
    	
    # 	# For track record
    #   mutate(
    #     SH_mu = (fit_SH[1] %$% estimate)[1], 
    #     SH_sigma = (fit_SH[1] %$% estimate)[2],
    #     SH_nu = (fit_SH[1] %$% estimate)[3]
    #     )
      
 	}) %>%
  collect() %>%
	ungroup() %>%
	gather(model, predicted, contains("predicted")) %>%
	ggplot(aes(x=`predicted` + 1, y=`read count normalised` + 1, label=symbol, color=model)) + 
	geom_abline(intercept = 0, slope = 1, color="grey") + 
	geom_point() + 
	facet_wrap(~ symbol, scales = "free")  +
	expand_limits(y=1, x=1) + 
	scale_y_log10() +
	scale_x_log10() +
	my_theme 

```



