---
title: "qq_plots_of_discrete_distributions_on_transcriptomic_data"
author: "Mangiola Stefano"
date: "31/01/2019"
output: html_document
---

```{r setup, include=FALSE, echo=FALSE}
# Import libraries

#setwd("/wehisan/bioinf/bioinf-data/Papenfuss_lab/projects/mangiola.s/PostDoc/RNAseq-noise-model")
set.seed(13254)

# Import libraries
library(tidyverse)
library(magrittr)
library(rstan)
library(tidybayes)
library(here)
library(foreach)
library(doParallel)
registerDoParallel()
# library(future)
# plan(multiprocess)

rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())

#devtools::load_all() # Sorry this costs me 30 minutes every day

my_theme = 	
	theme_bw() +
	theme(
		panel.border = element_blank(),
		axis.line = element_line(),
		panel.grid.major = element_line(size = 0.2),
		panel.grid.minor = element_line(size = 0.1),
		text = element_text(size=12),
		legend.position="bottom",
		aspect.ratio=1,
		axis.text.x = element_text(angle = 90, hjust = 1),
		strip.background = element_blank(),
		axis.title.x  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
		axis.title.y  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
	)

# Tools
source("https://gist.githubusercontent.com/stemangiola/90a528038b8c52b21f9cfa6bb186d583/raw/dbd92c49fb03fb05ab0b465704b99c0a39e654d5/transcription_tool_kit.R")
source("https://gist.githubusercontent.com/stemangiola/dd3573be22492fc03856cd2c53a755a9/raw/e4ec6a2348efc2f62b88f10b12e70f4c6273a10a/tidy_extensions.R")



```

## T cells most heavy-tailed genes

```{r load_data, cache=TRUE}
counts_noisy = read_csv("t_cells_with_error_qq_plot_means.csv")
```

```{r predict_values, cache=TRUE, fig.dim = c(9, 9)}

	counts_noisy %>%
  inner_join((.) %>% distinct(symbol, `gene error mean`) %>% arrange(`gene error mean` %>% desc) %>% head(n=2) %>% tail(n=1)) 
  arrange(`gene error mean` %>% desc) %>%
  mutate(symbol = factor(symbol, unique(symbol))) %>%

  #filter(symbol=="FOSB") %>%
  
  # Infer NB

	#do_parallel_start(40, "symbol") %>%
  group_by(symbol) %>%
   do({
 
     `%>%` = magrittr::`%>%`
		library(tidyverse)
		library(magrittr)
   PIG_model =
    rstan::stan_model(
    "/wehisan/bioinf/bioinf-data/Papenfuss_lab/projects/mangiola.s/PostDoc/RNAseq-noise-model/stan/poisson_inverseGaussian.stan" #  here::here("stan",sprintf("%s.stan", "poisson_inverseGaussian"))
    )

    PT_model =
      rstan::stan_model(
       "/wehisan/bioinf/bioinf-data/Papenfuss_lab/projects/mangiola.s/PostDoc/RNAseq-noise-model/stan/poisson_tweedie.stan" #" here::here("stan",sprintf("%s.stan", "poisson_tweedie"))
      )
browser()

     # lines(seq(0,1000, 10) , sapply(seq(0,1000, 10)  , gamlss.dist::dSICHEL,mu= 100, sigma=0.05, nu=-0.3, log=T))
    
    # Negative binomial
    fit_NB =  (.) %>%
       pull(`read count normalised`) %>%
      MASS::fitdistr("Negative Binomial")
    
    #PIG
    fit_PIG = 
      rstan::optimizing(
        PIG_model, 
        data = list(N = (.) %>% nrow, y = (.) %>% pull(`read count normalised`)),
        init = list(mu = (.) %>% pull(`read count normalised`) %>% `+` (1) %>% log %>% mean)
        ) %$% par
    
        #PIG
    fit_PT = 
      rstan::optimizing(
        PT_model, 
        data = list(N = (.) %>% nrow, y = (.) %>% pull(`read count normalised`)),
        init = list(mu = (.) %>% pull(`read count normalised`) %>% `+` (1) %>% log %>% mean)
        ) %$% par
    
    # # Sichel
    # fit_SH =  (.) %>%
    #    arrange(`read count`) %>%
    #    pull(`read count`) %>%
    #   MASS::fitdistr(gamlss.dist::dSICHEL, list(mu= 100, sigma=0.1, nu=-5), upper = c(4000, 20, 0),  lower=c(0, 0.0001, -10))

    qPT = function(pp, mu, sigma, nu){
      tweeDEseq::rPT(100000, mu = mu, D =sigma, a = nu, max=10000) %>% 
      sort %>% 
      `[` (
          pp %>% 
            multiply_by(100000) %>% 
            round
        )
    }
    
    (.) %>%
      arrange(`read count normalised`) %>%
      mutate(
      	predicted_NB = 
      		qnbinom(
	      		ppoints(`read count normalised`), 
	      		size=(fit_NB[1] %$% estimate)[1], 
	      		mu=(fit_NB[1] %$% estimate)[2]
	      		)
      ) %>%
      # PIG
      mutate(
      	predicted_PIG = 
      		gamlss.dist::qPIG(
      			ppoints(`read count normalised`), 
      			mu = fit_PIG[2] %>% exp, 
      			sigma=fit_PIG[1]
      			
      	) 
      ) %>%
      mutate(
      	predicted_PT = 
      		qPT(
      			ppoints(`read count normalised`), 
      			mu = fit_PT[1] %>% exp, 
      			sigma = fit_PT[2], 
      			nu = fit_PT[3]
      	) 
      ) 
    #%>%
      # mutate(
      # 	predicted_SH = 
      # 		gamlss.dist::qSICHEL(
      # 			ppoints(`read count`), 
      # 			nu = (fit_SH[1] %$% estimate)[3], 
      # 			sigma=(fit_SH[1] %$% estimate)[2], 
      # 			mu=(fit_SH[1] %$% estimate)[1], 
      # 			max.value = max(`read count`)
      # 			)
      # 	) %>%
    	
    # 	# For track record
    #   mutate(
    #     SH_mu = (fit_SH[1] %$% estimate)[1], 
    #     SH_sigma = (fit_SH[1] %$% estimate)[2],
    #     SH_nu = (fit_SH[1] %$% estimate)[3]
    #     )
      
 	}) %>%
	do_parallel_end() %>%
	ungroup() %>%
	gather(model, predicted, contains("predicted")) %>%
	ggplot(aes(x=`predicted` + 1, y=`read count normalised` + 1, label=symbol, color=model)) + 
	geom_abline(intercept = 0, slope = 1, color="grey") + 
	geom_point() + 
	facet_wrap(~ symbol, scales = "free")  +
	expand_limits(y=1, x=1) + 
	scale_y_log10() +
	scale_x_log10() +
	my_theme 

```

## 1000 genes plot ranks

```{r rank, cache=TRUE, fig.dim = c(9, 9)}

shards = 56
n_genes = 100
counts_source = "Acute_Myeloid_Leukemia_Primary_Blood_Derived_Cancer_-_Peripheral_Blood"

xx =
  
  read_csv(here(
    "big_data",
    "tibble_TCGA_files",
    paste0(counts_source, ".csv")
  )) %>%
  mutate(symbol = ens_iso) %>%
  
  # Median redundant
  do_parallel_start(40, "symbol") %>%
  do({
    `%>%` = magrittr::`%>%`
    library(tidyverse)
    library(magrittr)
    
    (.) %>%
      group_by(sample, symbol) %>%
      summarise(`read count` = `read count` %>% median(na.rm = T)) %>%
      ungroup()
  }) %>%
  do_parallel_end() %>%
  
  # Normalise
  norm_RNAseq(
    sample_column = "sample",
    gene_column = "symbol",
    value_column = "read count",
    cpm_threshold = 10,
    prop = 9 / 10
  ) %>%
  mutate(`read count normalised` = `read count normalised` %>% as.integer) %>%
  mutate(`read count normalised log` = `read count normalised` %>% `+` (1) %>% log) %>%
  
  # hOUSEKEEPING
  mutate(symbol = symbol %>% as.factor) %>%
  mutate(sample = sample %>% as.factor) %>%
  
  #
  filter(!filt_for_calc) %>%
  
  inner_join((.) %>% distinct(symbol) %>% sample_n(n_genes)) %>%
  
  # Add sample idx
  mutate(sample_idx = as.integer(factor(sample))) %>%
  left_join((.) %>% group_by(symbol) %>% summarise(s = `read count normalised` %>% sum) %>% arrange(s)) %>%
  filter(s > 0) %>%
  do_parallel_start(40, "symbol") %>%
  do({
    `%>%` = magrittr::`%>%`
    library(tidyverse)
    library(magrittr)
    library(rstan)
    PIG_model =
  rstan::stan_model(
    here::here("stan",sprintf("%s.stan", "poisson_inverseGaussian"))
  )
    
    (.) %>% group_by(symbol) %>%
      do({
        `%>%` = magrittr::`%>%`
        # library(tidyverse)
        # library(magrittr)
        # lines(seq(0,1000, 10) , sapply(seq(0,1000, 10)  , gamlss.dist::dSICHEL,mu= 100, sigma=0.05, nu=-0.3, log=T))
        #browser()
        # Negative binomial
        fit_NB =  (.) %>%
          pull(`read count normalised`) %>%
          MASS::fitdistr("Negative Binomial")
        
        #PIG
        fit_PIG =
          rstan::optimizing(
            PIG_model,
            data = list(
              N = (.) %>% nrow,
              y = (.) %>% pull(`read count normalised`)
            ),
            init = list(
              mu = (.) %>% pull(`read count normalised`) %>% `+` (1) %>% log %>% mean
            )
          ) %$% par
        
        # # Sichel
        # fit_SH =  (.) %>%
        #    arrange(`read count`) %>%
        #    pull(`read count`) %>%
        #   MASS::fitdistr(gamlss.dist::dSICHEL, list(mu= 100, sigma=0.1, nu=-5), upper = c(4000, 20, 0),  lower=c(0, 0.0001, -10))
        
        
        (.) %>%
          arrange(`read count normalised`) %>%
          mutate(predicted_NB =
                   qnbinom(
                     ppoints(`read count normalised`),
                     size = (fit_NB[1] %$% estimate)[1],
                     mu = (fit_NB[1] %$% estimate)[2]
                   )) %>%
          # PIG
          mutate(predicted_PIG =
                   gamlss.dist::qPIG(
                     ppoints(`read count normalised`),
                     mu = fit_PIG[2] %>% exp,
                     sigma = fit_PIG[1]
                     
                   ))
        #%>%
        # mutate(
        # 	predicted_SH =
        # 		gamlss.dist::qSICHEL(
        # 			ppoints(`read count`),
        # 			nu = (fit_SH[1] %$% estimate)[3],
        # 			sigma=(fit_SH[1] %$% estimate)[2],
        # 			mu=(fit_SH[1] %$% estimate)[1],
        # 			max.value = max(`read count`)
        # 			)
        # 	) %>%
        
        # 	# For track record
        #   mutate(
        #     SH_mu = (fit_SH[1] %$% estimate)[1],
        #     SH_sigma = (fit_SH[1] %$% estimate)[2],
        #     SH_nu = (fit_SH[1] %$% estimate)[3]
        #     )
        
      }) %>% ungroup()
  }) %>%
  do_parallel_end() %>%
  gather(model, predicted, contains("predicted")) %>%
  mutate(`log of error` = (`read count normalised` - predicted) %>% abs %>% `+` (1) %>% log) %>%
  mutate(`error of log` = (log(`read count normalised` + 1) - log(predicted + 1)) %>% abs) %>%
  mutate(`error scaled` =  ((`read count normalised` - predicted) / (`read count normalised` + 1)
  ) %>%  abs) %>%
  left_join((.) %>%
              group_by(symbol, model) %>%
              summarise(`gene error mean` = `error of log` %>% mean))
```

```{r, fig.dim = c(9, 9)}
xx %>%
  distinct(symbol, model, `gene error mean`) %>%
  arrange(`gene error mean` %>% desc) %>%
  mutate(symbol = factor(symbol, unique(symbol))) %>%
  ggplot(aes(x = symbol, y = `gene error mean`, color = model)) + geom_point() + my_theme

xx %>% 
  filter(symbol %in% c("ENSG00000242574.7",  "ENSG00000256338.2", "ENSG00000151612.14",  "ENSG00000137460.7")) %>%
	ggplot(aes(x=`predicted` + 1, y=`read count normalised` + 1, label=symbol, color=model)) + 
	geom_abline(intercept = 0, slope = 1, color="grey") + 
	geom_point() + 
	facet_wrap(~ symbol, scales = "free")  +
	expand_limits(y=1, x=1) + 
	scale_y_log10() +
	scale_x_log10() +
	my_theme 

```


